cassandra_test.go

```diff
@@ -312,6 +313,123 @@ func TestPaging(t *testing.T) {
 	}
 }
 
+func TestPagingWithAllowFiltering(t *testing.T) {
+	session := createSession(t)
+
+	t.Cleanup(func() {
+		if err := session.Query("DROP TABLE gocql_test.pagging_with_allow_filtering").Exec(); err != nil {
+			t.Fatal("drop table:", err)
+		}
+		session.Close()
+	})
+
+	if session.cfg.ProtoVersion == 1 {
+		t.Skip("Paging not supported. Please use Cassandra >= 2.0")
+	}
+
+	const (
+		targetP1      = 50
+		targetP2      = 50
+		expectedCount = 30
+	)
+
+	paginatedSelect := fmt.Sprintf("SELECT c1, f1 FROM gocql_test.pagging_with_allow_filtering WHERE p1 = %d AND p2 = %d AND f1 < %d ALLOW FILTERING;", targetP1, targetP2, expectedCount)
+	validateResult := func(t *testing.T, results []int) {
+		if len(results) != expectedCount {
+			t.Fatalf("expected 100, got %d: %d", len(results), results)
+		}
+
+		sort.Ints(results)
+
+		expect := make([]int, 0, expectedCount)
+		for i := 0; i < expectedCount; i++ {
+			expect = append(expect, i)
+		}
+
+		if !reflect.DeepEqual(results, expect) {
+			t.Fatalf("expected %v\ngot %v", expect, results)
+		}
+	}
+
+	t.Run("Prepare", func(t *testing.T) {
+		if err := createTable(session,
+			"CREATE TABLE gocql_test.pagging_with_allow_filtering (p1 int, p2 int, c1 int, f1 int, "+
+				"PRIMARY KEY ((p1, p2), c1)) WITH CLUSTERING ORDER BY (c1 DESC)"); err != nil {
+			t.Fatal("create table:", err)
+		}
+
+		// Insert extra records
+		for i := 0; i < 100; i++ {
+			if err := session.Query("INSERT INTO gocql_test.pagging_with_allow_filtering (p1,p2,c1,f1) VALUES (?,?,?,?)", i, i, i, i).Exec(); err != nil {
+				t.Fatal("insert:", err)
+			}
+		}
+
+		// Insert records to a target partition
+		for i := 0; i < 100; i++ {
+			if err := session.Query("INSERT INTO gocql_test.pagging_with_allow_filtering (p1,p2,c1,f1) VALUES (?,?,?,?)", targetP1, targetP2, i, i).Exec(); err != nil {
+				t.Fatal("insert:", err)
+			}
```

Comment:
Maybe worth also to delete some rows in the middle of the range, so we will have a range tombstone and we will cover this as well.
We had several issues with pagination when there are empty pages in the middle.

Reply:
done, take a look.